### PR TITLE
api dependency reduction - authentication

### PR DESCRIPTION
**Pull request by Joerger**
`lib/types/authentication.go` was originally moved to `api/types` to simplify the first `lib/services` refactor. Now it is clear that this should have stayed in `lib/services`, and just needed some minor refactoring to be moved back without errors. This also removes several dependencies in the api.

Changes:
 - Moved `user.Equals` method and some user validation to `lib/services/user.go`.
 - Converted `U2FRegistrationData` and `LocalAuthSecrets` methods to functions so that they can be defined in `lib/services`.

Needs to be rebased to master after #5384 is merged

### CODE REVIEW

File: lib/services/authentication.go [hunk 1]
```diff
@@ -526,14 +507,10 @@ var authPreferenceMarshaler AuthPreferenceMarshaler = &teleportAuthPreferenceMar
 
 // SetAuthPreferenceMarshaler sets global AuthPreferenceMarshaler
 func SetAuthPreferenceMarshaler(m AuthPreferenceMarshaler) {
-	marshalerMutex.Lock()
-	defer marshalerMutex.Unlock()
 	authPreferenceMarshaler = m
```
Comment (a-palchikov) [inline lib/services/authentication.go:510]: Why this change - is it safe to assume the marshaller is only set from a single goroutine?

File: lib/services/user.go [hunk 2]
```diff
@@ -22,6 +22,54 @@ import (
 	"github.com/gravitational/trace"
 )
 
+// ValidateUser validates the User and sets default values
+func ValidateUser(u User) error {
```
Comment (a-palchikov) [inline lib/services/user.go:26]: I think it could be less friction to keep these methods on an aliased type? WDYT?
This will definitely help with names such as `LocalAuthSecretsEquals`.

Comment (Joerger) [reply lib/services/user.go:26]: I agree, this is something I've been talking about with others and could help differentiate our api types from our business logic / service types. For now the plan is to make the changes like this, but once it hits a critical mass and other significant tasks are complete, I want to refactor this.

File: lib/services/authentication.go (hunk 1 above)
Comment (Joerger) [reply lib/services/authentication.go:510]: The marshalerMutex is made irrelevant by #5384. I'm going to rebase off of #5384 and draft this PR for now.

File: lib/services/authentication.go [hunk 3]
```diff
@@ -19,13 +19,401 @@ limitations under the License.
...
 	"github.com/gravitational/teleport/lib/utils"
+
 	"github.com/gravitational/trace"
+	"github.com/pquerna/otp/totp"
+	"github.com/tstranex/u2f"
+	"golang.org/x/crypto/bcrypt"
 )
 
+// AuthPreference defines the authentication preferences for a specific
+// cluster. It defines the type (local, oidc) and second factor (off, otp, oidc).
```
Comment (a-palchikov) [inline lib/services/authentication.go:41]:
```suggestion
// cluster. It defines the type (local, oidc) and second factor (otp, oidc or off).
```

File: lib/services/authentication.go [hunk 4]
```diff
@@ -19,13 +19,401 @@ limitations under the License.
...
+
+	// CheckAndSetDefaults sets and default values and then
+	// verifies the constraints for AuthPreference.
+	CheckAndSetDefaults() error
+
+	// String represents a human readable version of authentication settings.
+	String() string
+}
+
+// NewAuthPreference is a convenience method to to create AuthPreferenceV2.
```
Comment (a-palchikov) [inline lib/services/authentication.go:78]: nit: when the implementation changes to instead construct a value of type with the next version, this comment will be obsolete.
```suggestion
// NewAuthPreference is a convenience method to to create an auth preference value
```

**Review by a-palchikov — APPROVED** [review]

File: lib/services/authentication.go [hunk 5]
```diff
@@ -19,13 +19,401 @@ limitations under the License.
...
+		},
+		Spec: AuthPreferenceSpecV2{
+			Type:         constants.Local,
+			SecondFactor: constants.OTP,
+		},
+	}
+}
+
+// AuthPreferenceV2 implements AuthPreference.
+type AuthPreferenceV2 struct {
```
Comment (russjones) [inline lib/services/authentication.go:113]: Why is this moving back?

Comment (Joerger) [reply lib/services/authentication.go:113]: I moved the entire file back to `lib/services` because it was no longer needed by `api/client`. It may need to be moved back once we convert the http endpoint to gRPC. My thought process during the dependency refactors has been to move everything that is not needed by the api back to services, now that I have more context and understanding of the code. I think I moved a lot of things that should have stayed in services.

**Review by russjones — APPROVED** [review]
